### PR TITLE
Address issues picked up by Dev15 code analysis

### PR DESCRIPTION
**Pull request by Penguinwizzard**
This is one of the things that was preventing us from moving our jenkins CI to dev15.

### CODE REVIEW

File: bin/ch/Debugger.cpp [hunk 1]
```diff
@@ -316,13 +316,13 @@ bool Debugger::SetBaseline()
 
     if(file != nullptr)
     {
-        int fileSize = _filelength(_fileno(file));
-        if (fileSize <= MAX_BASELINE_SIZE)
+        long fileSize = _filelength(_fileno(file));
+        if (0 <= fileSize && fileSize <= MAX_BASELINE_SIZE)
         {
-            script = new char[fileSize + 1];
+            script = new char[(size_t)fileSize + 1];
```
Comment (boingoing) [inline bin/ch/Debugger.cpp]: Is this required?

File: lib/Runtime/Language/SimdBool32x4OperationX86X64.cpp [hunk 2]
```diff
@@ -11,7 +11,7 @@ namespace Js
     SIMDValue SIMDBool32x4Operation::OpBool32x4(bool x, bool y, bool z, bool w)
     {
         X86SIMDValue x86Result;
-        x86Result.m128i_value = _mm_set_epi32(w * -1, z * -1, y * -1, x * -1);
+        x86Result.m128i_value = _mm_set_epi32((w?1:0) * -1, (z?1:0) * -1, (y?1:0) * -1, (x?1:0) * -1);
```
Comment (boingoing) [inline lib/Runtime/Language/SimdBool32x4OperationX86X64.cpp]: Could this be simpler? 
```c++
_mm_set_epi32(w?-1:0, z?-1:0, y?-1:0, x?-1:0)
```

File: lib/Runtime/Library/SharedArrayBuffer.h [hunk 3]
```diff
@@ -194,7 +194,7 @@ namespace Js
 
         Waiters * m_waiters;
 
-        // Below CS is used for synchronizig access in wait/wake API
+        // Below CS is used for synchronizing access in wait/wake API
```
Comment (boingoing) [inline lib/Runtime/Library/SharedArrayBuffer.h]: Prefast finds typos now?

File: lib/Runtime/Types/DictionaryTypeHandler.cpp [hunk 4]
```diff
@@ -1569,11 +1569,16 @@ namespace Js
...
                 {
                     *getter = instance->GetSlot(descriptor->GetGetterPropertyIndex());
+                    *setter = nullptr;
                     getset = true;
                 }
                 if (descriptor->GetSetterPropertyIndex() != NoSlots)
                 {
                     *setter = instance->GetSlot(descriptor->GetSetterPropertyIndex());
+                    if(!getset) {
+                        // if we didn't set the getter above, we need to set it here
```
Comment (boingoing) [inline lib/Runtime/Types/DictionaryTypeHandler.cpp]: `getter` is an out param, right? How can we get away without setting it on all paths here? Why not just assign it and `setter` as well to nullptr right at the top of this function and then we don't care if it was set in above conditional?

**Comment by boingoing** [comment]
By the way, should this make it into release/1.10 or hold on until after we cut that branch?

File: bin/ch/Debugger.cpp (hunk 1 above)
Comment (Penguinwizzard) [reply bin/ch/Debugger.cpp]: Array size declarations are size_t, and the analysis gets unhappy if we have a different type used in the addition here before the implicit conversion. This is notable for types shorter than size_t, as the overflow would occur before the conversion to size_t. I believe that on all currently-supported platforms this is indeed not required.

File: lib/Runtime/Language/SimdBool32x4OperationX86X64.cpp (hunk 2 above)
Comment (Penguinwizzard) [reply lib/Runtime/Language/SimdBool32x4OperationX86X64.cpp]: Yes, you're correct, I'll fix that in the next update.

File: lib/Runtime/Library/SharedArrayBuffer.h (hunk 3 above)
Comment (Penguinwizzard) [reply lib/Runtime/Library/SharedArrayBuffer.h]: That being said, there are tools that spell-check comments, but I doubt that we'd find much benefit from using them.

File: lib/Runtime/Types/DictionaryTypeHandler.cpp (hunk 4 above)
Comment (Penguinwizzard) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Due to the _Success_(return) annotation, the outparam annotations are only required to be valid if the function returns true. I agree that it's a bit poorly designed, but the primary goal here was to align things with existing annotations.

File: lib/Runtime/Library/SharedArrayBuffer.h (hunk 3 above)
Comment (boingoing) [reply lib/Runtime/Library/SharedArrayBuffer.h]: I was just taking the opportunity to make a joke. We have typos in comments all over but fixing them like this - as a one-off thing - is probably best.

File: lib/Runtime/Types/DictionaryTypeHandler.cpp (hunk 4 above)
Comment (boingoing) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Alright but I object to the silliness of that annotation.

File: lib/Runtime/Library/SharedArrayBuffer.h (hunk 3 above)
Comment (Penguinwizzard) [reply lib/Runtime/Library/SharedArrayBuffer.h]: Ah, sorry. 👍

**Comment by Penguinwizzard** [comment]
I'm leaning toward putting it in master after the 1.10 cut. The downside of that is that we may end up with changes to release/1.10 breaking the CI when pumped, but that shouldn't be too common.

File: lib/Runtime/Library/SharedArrayBuffer.h (hunk 3 above)
Comment (boingoing) [reply lib/Runtime/Library/SharedArrayBuffer.h]: np 👌

File: lib/Common/Core/EtwTraceCore.h [hunk 5]
```diff
@@ -74,12 +74,18 @@ CompileAssert(false)
 #define MCGEN_PRIVATE_ENABLE_CALLBACK_V2(SourceId, ControlCode, Level, MatchAnyKeyword, MatchAllKeyword, FilterData, CallbackContext) \
        EtwCallback(ControlCode, CallbackContext)
 
+// Work-around for a bug in the instrumentationevents generator
+#pragma prefast(push)
+#pragma prefast(disable:__WARNING_USING_UNINIT_VAR, "The ETW data generated from the manifest includes a default null function which uses unintialized memory.")
```
Comment (MSLaguana) [inline lib/Common/Core/EtwTraceCore.h]: Is this because of how our manifest is structured, or would any project using ETW hit this same warning?

File: lib/Common/DataStructures/BaseDictionary.h [hunk 6]
```diff
@@ -44,20 +44,29 @@ namespace JsDiag
...
     public:
-        void BeginResize() {}
-        void EndResize() {}
+#pragma prefast(suppress:__WARNING_FAILING_TO_ACQUIRE_MEDIUM_CONFIDENCE)
+        void _Acquires_lock_(cs.cs) BeginResize() {}
+#pragma prefast(suppress:__WARNING_FAILING_TO_RELEASE_MEDIUM_CONFIDENCE)
+        void _Releases_lock_(cs.cs) EndResize() {}
+    private:
+        // For prefast analysis, we need to have a somewhat similar shape for both locks
+        Field(contentStruct) cs;
```
Comment (MSLaguana) [inline lib/Common/DataStructures/BaseDictionary.h]: This is a bit of a bummer, having to waste some space for no real reason. I'm not super familiar with the semantics of SAL locking annotations; would it make sense to refer to `this` instead of the actual lock (or the fake lock here), and elsewhere treat the various lock classes as critical sections / things-that-get-locked?

File: lib/Common/PlatformAgnostic/DateTimeInternal.h [hunk 7]
```diff
@@ -61,7 +61,7 @@ namespace DateTime
         uint32 lastTimeZoneUpdateTickCount;
 
         void UpdateTimeZoneInfo();
-        UtilityPlatformData(): lastTimeZoneUpdateTickCount(0) { }
+        UtilityPlatformData() : lastTimeZoneUpdateTickCount(0) { GetTimeZoneInformation(&timeZoneInfo); }
```
Comment (MSLaguana) [inline lib/Common/PlatformAgnostic/DateTimeInternal.h]: Instead of directly calling `GetTimeZoneInformation` here, would it be acceptable to call `UpdateTimeZoneInfo` to do it instead?

File: lib/Runtime/Language/ModuleNamespace.h [hunk 8]
```diff
@@ -61,7 +61,7 @@ namespace Js
         virtual BOOL DeleteItem(uint32 index, PropertyOperationFlags flags) override { return true; }
         virtual BOOL GetEnumerator(JavascriptStaticEnumerator * enumerator, EnumeratorFlags flags, ScriptContext* requestContext, EnumeratorCache * enumeratorCache = nullptr);
         virtual BOOL SetAccessors(PropertyId propertyId, Var getter, Var setter, PropertyOperationFlags flags = PropertyOperation_None) override { return false; }
-        virtual BOOL GetAccessors(PropertyId propertyId, Var *getter, Var *setter, ScriptContext * requestContext) override { return false; }
+        virtual BOOL GetAccessors(PropertyId propertyId, __out Var *getter, __out Var *setter, ScriptContext * requestContext) override { getter = nullptr; setter = nullptr; return false; }
```
Comment (MSLaguana) [inline lib/Runtime/Language/ModuleNamespace.h]: Should this be `*getter = nullptr; *setter = nullptr`?

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 9]
```diff
@@ -2158,6 +2158,7 @@ namespace Js
                     ::Math::DefaultOverflowPolicy();
```
Comment (MSLaguana) [inline lib/Runtime/Library/JavascriptProxy.cpp]: Should this use `UInt16Math` instead?

File: lib/Runtime/Types/DictionaryTypeHandler.cpp (hunk 4 above)
Comment (MSLaguana) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Can we have a getter without a setter, or vice versa? I thought that internally we constructed them in pairs, even if one was a stub "I don't exist, why are you calling me" function.

File: bin/ch/RuntimeThreadData.cpp [hunk 10]
```diff
@@ -25,10 +25,17 @@ RuntimeThreadData::RuntimeThreadData()
     this->hevntInitialScriptCompleted = CreateEvent(NULL, TRUE, FALSE, NULL);
     this->hevntReceivedBroadcast = CreateEvent(NULL, FALSE, FALSE, NULL);
     this->hevntShutdown = CreateEvent(NULL, TRUE, FALSE, NULL);
+    this->hSemaphore = nullptr;
+    this->hThread = nullptr;
```
Comment (sigatrev) [inline bin/ch/RuntimeThreadData.cpp]: is there a reason not to do these in an initializer list, or better yet, in class initialization?

File: lib/Runtime/Library/SharedArrayBuffer.h (hunk 3 above)
Comment (dilijev) [reply lib/Runtime/Library/SharedArrayBuffer.h]: We (I) actually did use some spell checker tools before we open-sourced ChakraCore -- didn't catch everything and more typos have creeped in since then. I support drive-by clean up. :)

File: lib/Runtime/Types/DictionaryTypeHandler.cpp (hunk 4 above)
Comment (dilijev) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Since `if(!getset)` looks like new code, can we use the existing convention (and preferred by many) of putting a space in `if (`?

**Comment by dilijev** [comment]
> should this make it into release/1.10 or hold on until after we cut that branch?

I'd prefer before the cut (this change included in 1.10) so that 1.10 can have modern CI config enabled.

File: lib/Common/DataStructures/BaseDictionary.h (hunk 6 above)
Comment (dilijev) [reply lib/Common/DataStructures/BaseDictionary.h]: As for the memory argument, how common is this object? We tend not to worry about memory usage in objects that have very few instances.

Comment (MSLaguana) [reply lib/Common/DataStructures/BaseDictionary.h]: It's on most of our native Dictionary classes. I was also somewhat concerned that it might change the memory layout in a meaningful way, but there are some static asserts that check for that so it seems to be fine.

As for the number of native dictionaries we have, it's probably more like tens than tens of thousands, so the actual impact is probably small, but I don't have solid numbers to back that up at the moment.

I do agree that my comment here shouldn't block the PR.

File: lib/Runtime/Types/DictionaryTypeHandler.cpp (hunk 4 above)
Comment (MSLaguana) [reply lib/Runtime/Types/DictionaryTypeHandler.cpp]: Also put the `{` on a new line, while you're at it.

**Comment by dilijev** [comment]
For reference the payload we'll be merging to update CI so that modern is Dev15 and Legacy is Dev14 is in #4767 -- we'll retarget that to whichever branch this change lands in.

File: lib/Runtime/Language/ModuleNamespace.h (hunk 8 above)
Comment (dilijev) [reply lib/Runtime/Language/ModuleNamespace.h]: Yes, probably, and with a check on whether the pointers passed in are not nullptr before assigning to them.

File: bin/ch/RuntimeThreadData.cpp (hunk 10 above)
Comment (Penguinwizzard) [reply bin/ch/RuntimeThreadData.cpp]: Yeah, changed in next commit.

File: lib/Common/Core/EtwTraceCore.h (hunk 5 above)
Comment (Penguinwizzard) [reply lib/Common/Core/EtwTraceCore.h]: I'm not entirely sure. I wasn't able to see a good way to fix it by changing the manifest, so I believe that it may be across many projects using ETW in a similar manner.

**Comment by dilijev** [comment]
>  *  Copyright (c) 2012 Two Blue Cubes Ltd. All rights reserved.

What do we do about making changes to this file into permanent updates -- should we find the source that generated this file and PR back, or just say we need this update and we're not planning to regen this file?

---
Refers to: bin/External/catch.hpp:6 in 7a9b9ea. [](commit_id = 7a9b9eafeb0d3e2135636a419748ecfb5ef6389b, deletion_comment = False)

File: bin/NativeTests/JsRTApiTest.cpp [hunk 11]
```diff
@@ -1243,7 +1244,8 @@ namespace JsRTApiTest
             size_t length;
             REQUIRE(JsStringToPointer(nameValue, &name, &length) == JsNoError);
 
-            CHECK(length == 1);
+            REQUIRE(length == 1);
+#pragma prefast(suppress:__WARNING_MAYBE_UNINIT_VAR, "The require on the previous line should ensure that name[0] is initialized")
```
Comment (dilijev) [inline bin/NativeTests/JsRTApiTest.cpp]: >__WARNING_MAYBE_UNINIT_VAR [](start = 25, length = 26)

Did not all warnings have these fancy named macros? (There are some you're still suppressing or disabling by number)

File: bin/ch/HostConfigFlags.cpp [hunk 12]
```diff
@@ -96,6 +96,7 @@ void HostConfigFlags::RemoveArg(int& argc, _Inout_updates_to_(argc, argc) LPWSTR
     Assert(index >= 0 && index < argc);
     for (int i = index + 1; i < argc; ++i)
     {
+#pragma prefast(suppress:6385, "Operation is safe but PREfast is difficult to convince")
```
Comment (dilijev) [inline bin/ch/HostConfigFlags.cpp]: >6385 [](start = 25, length = 4)

suppress warning by name macro instead of number?

File: bin/ch/WScriptJsrt.cpp [hunk 13]
```diff
@@ -35,6 +35,7 @@
 #endif // FreeBSD or unix ?
 #endif // _WIN32 ?
 
+#pragma prefast(disable:26444, "This warning unfortunately raises false positives when auto is used for declaring the type of an iterator in a loop.")
```
Comment (dilijev) [inline bin/ch/WScriptJsrt.cpp]: >26444 [](start = 24, length = 5)

suppress by name macro instead of number?

**Comment by MSLaguana** [comment]
Regarding the catch.hpp change; looks like their upstream has that fix (or an equivalent fix at least): https://github.com/catchorg/Catch2/blob/master/single_include/catch.hpp#L10054

File: lib/Common/PlatformAgnostic/DateTimeInternal.h [hunk 14]
```diff
@@ -78,13 +78,14 @@ namespace DateTime
...
-            baseMsCount(0),
-            fHiResAvailable(true),
             dLastTime(0),
             dAdjustFactor(1),
-            fReset(true)
+            baseMsCount(0),
+            freq(0),
+            fReset(true),
+            fInit(false),
+            fHiResAvailable(true)
```
Comment (dilijev) [inline lib/Common/PlatformAgnostic/DateTimeInternal.h]: was there a warning about initialization order or is this just drive-by cleanup?

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 15]
```diff
@@ -2178,13 +2179,19 @@ namespace Js
                 calleeInfo.Flags = (CallFlags)(calleeInfo.Flags | CallFlags_ExtraArg | CallFlags_NewTarget);
             }
 
-            for (uint argCount = 0; argCount < args.Info.Count; argCount++)
+            for (ushort argCount = 0; argCount < (ushort)args.Info.Count; argCount++)
```
Comment (dilijev) [inline lib/Runtime/Library/JavascriptProxy.cpp]: >args.Info.Count [](start = 57, length = 15)

is it safe to cast args.Info.Count to ushort? If so, can we make args.Info.Count have the correctly-sized type?

File: lib/Runtime/Language/ModuleNamespace.h (hunk 8 above)
Comment (sethbrenith) [reply lib/Runtime/Language/ModuleNamespace.h]: Ideally these should have been statically verified as non-null because we said _Out_ and not _Out_opt_, right?

File: lib/Runtime/Library/JavascriptProxy.cpp [hunk 16]
```diff
@@ -2178,13 +2179,19 @@ namespace Js
...
             {
+                AnalysisAssert(newCount >= ((ushort)args.Info.Count));
+                AnalysisAssert(argCount < newCount);
+                AnalysisAssert(argCount < (ushort)args.Info.Count);
+                AnalysisAssert(sizeof(Var*) == sizeof(void*));
+                AnalysisAssert(sizeof(Var*) * argCount < sizeof(void*) * newCount);
+#pragma prefast(suppress:__WARNING_WRITE_OVERRUN, "This is a false positive, and all of the above analysis asserts still didn't convince prefast of that.")
```
Comment (sethbrenith) [inline lib/Runtime/Library/JavascriptProxy.cpp]: have we opened bugs for false positives?

File: lib/Runtime/Library/JavascriptProxy.cpp (hunk 15 above)
Comment (MSLaguana) [reply lib/Runtime/Library/JavascriptProxy.cpp]: Further up around https://github.com/Microsoft/ChakraCore/pull/5272/files#diff-771647fb11e03690664371970f34eeaaR2126 we check that the value fits in a ushort and throw an error if it doesn't. We do allow for larger argument counts in some circumstances though, so `args.Info.Count` is a uint.

File: lib/Runtime/Language/ModuleNamespace.h (hunk 8 above)
Comment (sethbrenith) [reply lib/Runtime/Language/ModuleNamespace.h]: Also, same pattern is in SpreadArgument.h

File: lib/Common/PlatformAgnostic/DateTimeInternal.h (hunk 14 above)
Comment (Penguinwizzard) [reply lib/Common/PlatformAgnostic/DateTimeInternal.h]: There's one in razzle prefast, and it helps for checking that there's no other uninitialized values when they're in the right order. We do not have initialization order warnings enabled in our jenkins runs.

File: bin/ch/HostConfigFlags.cpp (hunk 12 above)
Comment (Penguinwizzard) [reply bin/ch/HostConfigFlags.cpp]: Done in next set of commits.

File: bin/NativeTests/JsRTApiTest.cpp (hunk 11 above)
Comment (Penguinwizzard) [reply bin/NativeTests/JsRTApiTest.cpp]: Nope, not all of them do. There are a couple cases where I missed using the named macro, though, that I should be fixing up in the next set of commits.

File: bin/ch/WScriptJsrt.cpp (hunk 13 above)
Comment (Penguinwizzard) [reply bin/ch/WScriptJsrt.cpp]: No macro for this one, unfortunately.

File: lib/Runtime/Language/ModuleNamespace.h (hunk 8 above)
Comment (Penguinwizzard) [reply lib/Runtime/Language/ModuleNamespace.h]: `*getter = nullptr; *setter = nullptr` would indeed be more correct, but the next set of commits avoids the issue entirely by making the function `_Success_(return) _Check_return_`, and not assigning to either, as the result is always false.

File: lib/Runtime/Library/JavascriptProxy.cpp (hunk 16 above)
Comment (Penguinwizzard) [reply lib/Runtime/Library/JavascriptProxy.cpp]: Not yet; I need to determine where we'd open bugs for these, and what would constitute an actionable report.

File: bin/GCStress/WeightedTable.h [hunk 17]
```diff
@@ -13,14 +13,14 @@ template <class T>
 class WeightedTable
 {
 public:
-    WeightedTable() :
+    WeightedTable() noexcept :
```
Comment (dilijev) [inline bin/GCStress/WeightedTable.h]: >noexcept [](start = 20, length = 8)

Did you mean to add this?

Comment (Penguinwizzard) [reply bin/GCStress/WeightedTable.h]: Yeah, it was that or a longer statement to disable the warning for just that line. This took fewer keypresses, and won't complicate any future efforts to go over all the warnings we've disabled.
